<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Johnny Lee">
<meta name="dcterms.date" content="2023-01-01">

<title>Biomedical Data Science (MATH11111) Assignment 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="assignment2_files/libs/clipboard/clipboard.min.js"></script>
<script src="assignment2_files/libs/quarto-html/quarto.js"></script>
<script src="assignment2_files/libs/quarto-html/popper.min.js"></script>
<script src="assignment2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="assignment2_files/libs/quarto-html/anchor.min.js"></script>
<link href="assignment2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="assignment2_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="assignment2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="assignment2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="assignment2_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Biomedical Data Science (MATH11111) Assignment 2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Johnny Lee </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 1, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="due-on-wednesday-6th-of-april-2023-500pm" class="level2">
<h2 class="anchored" data-anchor-id="due-on-wednesday-6th-of-april-2023-500pm"><strong>Due on Wednesday, 6th of April 2023, 5:00pm</strong></h2>
<p>The assignment is marked out of 100 points, and will contribute to <strong><em>20%</em></strong> of your final mark. Please complete this assignment using <strong>rmarkdown/quarto file and knit/render this document in PDF format</strong> and submit using the <strong>gradescope link on Learn</strong>. If you cannot knit/render to PDF directly, knit/render it to word and you should be able to either convert to PDF or print it and scan to PDF using a scanning app on your phone. If you have any code that does not run you will be able to knit/render the document so comment it as you might still get some grades for partial code.</p>
<p><strong>Clear and reusable code will be rewarded so pay attention to indentation, choice of variable identifiers, comments, error checking, etc.</strong> An initial code chunk is provided after each subquestion but <strong>create as many chunks as you feel is necessary</strong> to make a clear report. Add plain text explanations in between the chunks as and when required and any comments necessary within code chunks to make it easier to follow your code/reasoning.</p>
</section>
<section id="problem-1-27-points" class="level2">
<h2 class="anchored" data-anchor-id="problem-1-27-points">Problem 1 (27 points)</h2>
<p>File <strong><em>wdbc2.csv</em></strong> (available from the accompanying zip folder on Learn) refers to a study of breast cancer where the outcome of interest is the type of the tumour (benign or malignant, recorded in column <code>diagnosis</code>). The study collected 30 imaging biomarkers on 569 patients.</p>
<section id="problem-1.a-7-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.a-7-points">Problem 1.a (7 points)</h3>
<p>Using package <code>caret</code>, create a data partition so that the training set contains 70% of the observations (set the random seed to 984065 beforehand). Fit both a ridge regression model and a lasso model which uses cross validation on the training set to diagnose the type of tumour from the 30 biomarkers. Then use a plot to help identify the penalty parameter <span class="math inline">\(\lambda\)</span> that maximizes the AUC. Note: There is no need to use the <code>prepare.glmnet()</code> function from lab 4, using <code>as.matrix()</code> with the required columns is sufficient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-1.b-2-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.b-2-points">Problem 1.b (2 points)</h3>
<p>Create a data table that for each value of <code>lambda.min</code> and <code>lambda.1se</code> for each model fitted in <strong>problem 1.a</strong> reports: * the corresponding AUC, * the corresponding model size. Use 3 significant figures for floating point values and comment on these results. <!-- Hint: The AUC values are stored in the field called `cvm`. --></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-1.c-7-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.c-7-points">Problem 1.c (7 points)</h3>
<p>Perform both backward (we denote this as <strong>model B</strong>) and forward (<strong>model S</strong>) stepwise selection on the same training set derived in <strong>problem 1.a</strong>. Report the variables selected and their standardised regression coefficients in inccreasing order of the absolute value of their standardised regression coefficient. Discuss the results and how the different variables entering or leaving the model influenced the final result.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-1.d-3-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.d-3-points">Problem 1.d (3 points)</h3>
<p>Compare the goodness of fit of <strong>model B</strong> and <strong>model S</strong> in an appropriate way.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-1.e-2-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.e-2-points">Problem 1.e (2 points)</h3>
<p>Compute the ROC curve and AUC value of the trained model for both <strong>model B</strong> and <strong>model S</strong>. Display with clear title, label and legend.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-1.f-6-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-1.f-6-points">Problem 1.f (6 points)</h3>
<p>Use the four models to predict the outcome for the observations in the test set (use the <span class="math inline">\(\lambda\)</span> at 1 standard error for the penalised models). Plot the ROC curves of these models (on the sameplot, using different colours) and report their test AUCs. Compare the training AUCs obtained in <strong>problems 1.b and 1.e</strong> with the test AUCs and discuss the fit of the different models. Display with clear title, label and legend.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="problem-2-40-points" class="level2">
<h2 class="anchored" data-anchor-id="problem-2-40-points">Problem 2 (40 points)</h2>
<p>File <strong><em>GDM.raw.txt</em></strong> (available from the accompanying zip folder on Learn) contains 176 <code>SNP</code>s to be studied for association with incidence of gestational diabetes (a form of diabetes that is specific to pregnant women). SNP names are given in the form <code>rs1234_X</code> where <code>rs1234</code> is the official identifier (rsID), and <code>X</code> (one of A, C, G, T) is the reference allele.</p>
<section id="problem-2.a-3-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.a-3-points">Problem 2.a (3 points)</h3>
<p>Read in file <strong><em>GDM.raw.txt</em></strong> into a data table named <code>gdm.dt</code>. Impute missing values in <code>gdm.dt</code> according to <code>SNP</code>-wise median allele count.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.b-8-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.b-8-points">Problem 2.b (8 points)</h3>
<p>Write function <code>univ.glm.test()</code> where it takes 3 arguements, <code>x</code>, <code>y</code> and <code>order</code>. <code>x</code> is a data table of <code>SNP</code>s, <code>y</code> is a binary outcome vector, and <code>order</code> is a boolean which takes <code>false</code> as a default value. The function should fit a logistic regression model for each <code>SNP</code> in <code>x</code>, and return a data table containing SNP names, regression coefficients, odds ratios, standard errors and p-values. If order is set to <code>TRUE</code>, the output data table should be ordered by increasing p-value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.c-5-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.c-5-points">Problem 2.c (5 points)</h3>
<p>Using function univ.glm.test(), run an association study for all the <code>SNP</code>s in <code>gdm.dt</code> against having gestational diabetes (column <code>pheno</code>). For the <code>SNP</code> that is most strongly associated to increased risk of gestational diabetes and the one with most significant protective effect, report the summary statistics from the GWAS as well as the 95% and 99% confidence intervals on the odds ratio.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.d-4points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.d-4points">Problem 2.d (4points)</h3>
<p>Merge your GWAS results with the table of gene names provided in file <strong><em>GDM.annot.txt</em></strong> (available from the accompanying zip folder on Learn). For <code>SNP</code>s that have p-value <span class="math inline">\(&lt; 10^{-4}\)</span> (<code>hit SNP</code>s) report <code>SNP</code> name, effect allele, chromosome number and corresponding gene name. Separately, report for each <code>hit SNP</code> the names of the genes that are within a 1Mb window from the <code>SNP</code> position on the chromosome. Note: That’s genes that fall within +/- 1,000,000 positions using the <code>pos</code> column in the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.e-8-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.e-8-points">Problem 2.e (8 points)</h3>
<p>Build a weighted genetic risk score that includes all <code>SNP</code>s with p-value <span class="math inline">\(&lt; 10^{-4}\)</span>, a score with all <code>SNP</code>s with p-value <span class="math inline">\(&lt; 10^{-3}\)</span>, and a score that only includes <code>SNP</code>s on the FTO gene (hint: ensure that the ordering of <code>SNP</code>s is respected). Add the three scores as columns to the <code>gdm.dt</code> data table. Fit the three scores in separate logistic regression models to test their association with gestational diabetes, and for each report odds ratio, 95% confidence interval and p-value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.f-4-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.f-4-points">Problem 2.f (4 points)</h3>
<p>File <strong><em>GDM.test.txt</em></strong> (available from the accompanying zip folder on Learn) contains genotypes of another 40 pregnant women with and without gestational diabetes (assume that the reference allele is the same one that was specified in file <strong><em>GDM.raw.txt</em></strong>). Read the file into variable <code>gdm.test</code>. For the set of patients in <code>gdm.test</code>, compute the three genetic risk scores as defined in <strong>problem 2.e</strong> using the same set of <code>SNP</code>s and corresponding weights. Add the three scores as columns to <code>gdm.test</code> (hint: use the same columnnames as before).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.g-4-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.g-4-points">Problem 2.g (4 points)</h3>
<p>Use the logistic regression models fitted in <strong>problem 2.e</strong> to predict the outcome of patients in <code>gdm.test</code>. Compute the test log-likelihood for the predicted probabilities from the three genetic risk score models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-2.h-4points" class="level3">
<h3 class="anchored" data-anchor-id="problem-2.h-4points">Problem 2.h (4points)</h3>
<p>File <strong><em>GDM.study2.txt</em></strong> (available from the accompanying zip folder on Learn) contains the summary statistics from a different study on the same set of <code>SNP</code>s. Perform a meta-analysis with the results obtained in <strong>problem 2.c</strong> (hint: remember that the effect alleles should correspond) and produce a summary of the meta-analysis results for the set of <code>SNP</code>s with meta-analysis p-value <span class="math inline">\(&lt; 10^{-4}\)</span> sorted by increasing p-value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="problem-3-33-points" class="level2">
<h2 class="anchored" data-anchor-id="problem-3-33-points">Problem 3 (33 points)</h2>
<p>File <strong><em>nki.csv</em></strong> (available from the accompanying zip folder on Learn) contains data for 144 breast cancer patients. The dataset contains a binary outcome variable (<code>Event</code>, indicating the insurgence of further complications after operation), covariates describing the tumour and the age of the patient, and gene expressions for 70 genes found to be prognostic of survival.</p>
<section id="problem-3.a-6-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-3.a-6-points">Problem 3.a (6 points)</h3>
<p>Compute the correlation matrix between the gene expression variables, and display it so that a block structure is highlighted using <code>corrplot</code> package. Discuss what you observe. Write some code to identify the unique pairs of (distinct) variables that have correlation coefficient greater than 0.80 in absolute value and report their correlation coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-3.b-8-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-3.b-8-points">Problem 3.b (8 points)</h3>
<p>Perform PCA analysis (only over the columns containing gene expressions), in order to derive a patient-wise summary of all gene expressions (dimensionality reduction). Decide which components to keep and justify your decision. Test if those principal components are associated with the outcome in unadjusted logistic regression models and in models adjusted for age, estrogen receptor and grade. Justify the difference in results between unadjusted and adjusted models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-3.c-8-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-3.c-8-points">Problem 3.c (8 points)</h3>
<p>Use plots to compare with the correlation structure observed in <strong>problem 2.a</strong> and to examine how well the dataset may explain your outcome. Discuss your findings and suggest any further steps if needed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="problem-3.d-11-points" class="level3">
<h3 class="anchored" data-anchor-id="problem-3.d-11-points">Problem 3.d (11 points)</h3>
<p>Based on the models we examined in the labs, fit an appropriate model with the aim to provide the most accurate prognosis you can for patients. Discuss and justify your decisions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>