<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Johnny MyungWon Lee (Modified from Debby Lipschutz &amp; Stuart McGurnaghan)">
<meta name="dcterms.date" content="2023-03-03">
<title>Lab 5 : Using Genetic Data to Predict Outcomes</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/clipboard/clipboard.min.js"></script>
<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/quarto.js"></script>
<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/popper.min.js"></script>
<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/anchor.min.js"></script>
<link href="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/kePrint-0.0.1/kePrint.js"></script><link href="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#gwas-analyses" id="toc-gwas-analyses" class="nav-link active" data-scroll-target="#gwas-analyses">GWAS analyses</a></li>
  <li><a href="#genetic-risk-scores" id="toc-genetic-risk-scores" class="nav-link" data-scroll-target="#genetic-risk-scores">Genetic Risk Scores</a></li>
  <li><a href="#predicting-with-scores" id="toc-predicting-with-scores" class="nav-link" data-scroll-target="#predicting-with-scores">Predicting with scores</a></li>
  <li><a href="#gwas-meta-analysis" id="toc-gwas-meta-analysis" class="nav-link" data-scroll-target="#gwas-meta-analysis">GWAS meta-analysis</a></li>
  <li>
<a href="#practical-exercises" id="toc-practical-exercises" class="nav-link" data-scroll-target="#practical-exercises">Practical Exercises</a>
  <ul class="collapse">
<li><a href="#question-1." id="toc-question-1." class="nav-link" data-scroll-target="#question-1.">Question 1.</a></li>
  <li><a href="#question-2" id="toc-question-2" class="nav-link" data-scroll-target="#question-2">Question 2</a></li>
  </ul>
</li>
  </ul></nav>
</div>
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Lab 5 : Using Genetic Data to Predict Outcomes</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Johnny MyungWon Lee (Modified from Debby Lipschutz &amp; Stuart McGurnaghan) </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 3, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header><section id="gwas-analyses" class="level1"><h1>GWAS analyses</h1>
<p>This section does not rely on any previous knowledge of genetics.</p>
<p>Given the size of the datasets and the large number of computations, usually genome-wide association studies are run with specialised software that is optimised to read data in binary format and can represent genetic data in memory efficiently (<code>plink</code>, <code>SNPTEST</code>, <code>RegScan</code>, etc). R is very inefficient memory-wise, and performs poorly in loops (that’s a reason why using vectorised code is always to be preferred). However we can still perform the same analyses in R, provided that datasets are not too large.</p>
<p><code>SNP</code>’s are locations on the human genome that differ in individuals across the population (each <code>SNP</code> on each of the <span class="math inline">\(23\)</span> pairs of chromosomes in humans has a <code>SNP</code> identifier for the purposes of these labs a position). Each <code>SNP</code> has two possible alleles and when we deal with the SNP data, we deal with one of the two. This is called the <code>reference</code> allele and each <code>SNP</code> will have the value <span class="math inline">\(0\)</span>, <span class="math inline">\(1\)</span>, or <span class="math inline">\(2\)</span>. assigned. Note that the reference allele may or may not be the minor allele. This means that the minority of the population is expected to have that allele. The reference alleles are set by looking at sample of the population who have donated their DNA to a study. An example of this would be a <code>SNP</code> that has a reference allele of <code>T</code> e.g.&nbsp;<code>rs4506565_T</code> and a person is assigned a value of <span class="math inline">\(2\)</span> for that <code>SNP</code> (opposed to <span class="math inline">\(0\)</span> or <span class="math inline">\(1\)</span>), then that person would have two copies of <code>T</code> reference allele (one from each parent). If the alternate allele in the population was <code>C</code>, this person would have no copies of the alternate. The association test is performed on the groups of people with and without the reference allele and the phenotype (observable characteristic or trait) of interest.</p>
<p>The simplest types of association studies rely on testing one <code>SNP</code> at a time for association with the outcome: for a continuous trait, we use linear regression, for a binary trait we use logistic regression.</p>
<p>In real analyses, it is common to include other relevant covariates in the models, to ensure that the <code>SNP</code> association is independent of those covariates.In particular, models are adjusted by an indicator variable corresponding to the cohort of origin (as they may have been genotyped with different technologies, or may have differences in the way their phenotypes are measured) and by the first few principal components (often <span class="math inline">\(5\)</span>-<span class="math inline">\(10\)</span> PCs) to remove possible confounding effects due to population stratification. It is also increasingly common to use linear mixed models to account for population structure and relatedness between individuals.</p>
<p>In what follows, we will perform <em>crude associations</em>, that is not adjusted for other covariates, bearing in mind that it is a simplification. We start by reading file <code>chrom21.csv</code> (available on Learn), which contains <span class="math inline">\(1500\)</span> <code>SNP</code>s from chromosome <span class="math inline">\(21\)</span> as well as a column with a quantitative outcome (the first column).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>chrom21 <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/chrom21.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Also genetic data can contain missing values, due to genotyping errors or poor quality reads that are discarded at the quality control step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">kable</span>(<span class="fu">table</span>(<span class="fu">is.na</span>(chrom21)), <span class="at">caption =</span> <span class="st">"Missing Values"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Missing Values</caption>
 <thead><tr>
<th style="text-align:left;"> Var1 </th>
   <th style="text-align:right;"> Freq </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> FALSE </td>
   <td style="text-align:right;"> 301447 </td>
  </tr>
<tr>
<td style="text-align:left;"> TRUE </td>
   <td style="text-align:right;"> 254 </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>A common approach to impute missing values in a <code>SNP</code> is by replacing them with the mean of the <code>SNP</code>: this corresponds to assign the expected allele dosages to the missing entries.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="cf">for</span> (colnm <span class="cf">in</span> <span class="fu">colnames</span>(chrom21[, <span class="sc">-</span><span class="dv">1</span>])) {</span>
<span id="cb3-2"><a href="#cb3-2"></a>  chrom21[[colnm]][<span class="fu">is.na</span>(chrom21[[colnm]])] <span class="ot">&lt;-</span> <span class="fu">mean</span>(chrom21[[colnm]], <span class="at">na.rm =</span> T)</span>
<span id="cb3-3"><a href="#cb3-3"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Given that we have a continuous outcome, to test for a crude association of a <code>SNP</code> it is sufficient to run a simple linear regression model, then check the summary statistics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>snp1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(outcome <span class="sc">~</span> snp1, <span class="at">data =</span> chrom21)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="fu">kable</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(snp1)), <span class="at">caption =</span> <span class="st">"Summary Statistics"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Summary Statistics</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> Estimate </th>
   <th style="text-align:right;"> Std. Error </th>
   <th style="text-align:right;"> t value </th>
   <th style="text-align:right;"> Pr(&gt;|t|) </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> (Intercept) </td>
   <td style="text-align:right;"> -0.2169840 </td>
   <td style="text-align:right;"> 0.0869401 </td>
   <td style="text-align:right;"> -2.4957863 </td>
   <td style="text-align:right;"> 0.0133805 </td>
  </tr>
<tr>
<td style="text-align:left;"> snp1 </td>
   <td style="text-align:right;"> -0.1410125 </td>
   <td style="text-align:right;"> 0.2689730 </td>
   <td style="text-align:right;"> -0.5242626 </td>
   <td style="text-align:right;"> 0.6006794 </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>This specific <code>SNP</code> is not associated with the outcome.</p>
<p>It is clear that we do not want to explicitly formulate a model for each <code>SNP</code> in the dataset, but want to create a small function that could be used within a <code>for</code> loop to automate the process.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># run univariate tests of associations for all SNPs(columns of az) </span></span>
<span id="cb5-2"><a href="#cb5-2"></a>univ.lm.test <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y) { </span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="fu">stopifnot</span> (<span class="fu">length</span>(x) <span class="sc">==</span> <span class="fu">length</span>(y)) </span>
<span id="cb5-4"><a href="#cb5-4"></a>  regr <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> x) </span>
<span id="cb5-5"><a href="#cb5-5"></a>  <span class="co"># remove the row corresponding to the intercept and the column containing </span></span>
<span id="cb5-6"><a href="#cb5-6"></a>  <span class="co"># the t-value, then convert to a data table </span></span>
<span id="cb5-7"><a href="#cb5-7"></a>  output <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(regr)))[<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="dv">3</span>]</span>
<span id="cb5-8"><a href="#cb5-8"></a>  <span class="co"># assign better column names </span></span>
<span id="cb5-9"><a href="#cb5-9"></a>  <span class="fu">colnames</span>(output) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"beta"</span>, <span class="st">"std.error"</span>, <span class="st">"p.value"</span>)</span>
<span id="cb5-10"><a href="#cb5-10"></a>  output</span>
<span id="cb5-11"><a href="#cb5-11"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With this function we can run the same model as before by running the following command.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">kable</span>(<span class="fu">univ.lm.test</span>(chrom21<span class="sc">$</span>snp1, chrom21<span class="sc">$</span>outcome), </span>
<span id="cb6-2"><a href="#cb6-2"></a>      <span class="at">caption =</span> <span class="st">"Summary Statistics using Function"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Summary Statistics using Function</caption>
 <thead><tr>
<th style="text-align:right;"> beta </th>
   <th style="text-align:right;"> std.error </th>
   <th style="text-align:right;"> p.value </th>
  </tr></thead>
<tbody><tr>
<td style="text-align:right;"> -0.1410125 </td>
   <td style="text-align:right;"> 0.268973 </td>
   <td style="text-align:right;"> 0.6006794 </td>
  </tr></tbody>
</table>
</div>
</div>
<p>Now this can be automated in a loop to run the association test over all <code>SNP</code>s: this may take several seconds, depending on the processor speed (as all models are independent, the process can be parallelised).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>crude <span class="ot">&lt;-</span> <span class="cn">NULL</span> </span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="cf">for</span> (snp <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">ncol</span>(chrom21)) {</span>
<span id="cb7-3"><a href="#cb7-3"></a>  <span class="co"># skip column 1 as it contains the outcome </span></span>
<span id="cb7-4"><a href="#cb7-4"></a>  crude <span class="ot">&lt;-</span> <span class="fu">rbind</span>(crude, <span class="fu">univ.lm.test</span>(chrom21[[snp]], </span>
<span id="cb7-5"><a href="#cb7-5"></a>                  chrom21<span class="sc">$</span>outcome) ) </span>
<span id="cb7-6"><a href="#cb7-6"></a>  } </span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co"># add column of snp identifiers </span></span>
<span id="cb7-8"><a href="#cb7-8"></a>crude <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">snp =</span> <span class="fu">colnames</span>(chrom21)[<span class="sc">-</span><span class="dv">1</span>], crude)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These associations can be visualised through a volcano plot, where strongest associations are those that are towards the top (small p-values) and further to the sides (large effect sizes) of the plot. For reference we also draw a horizontal line corresponding to the standard genome-wide significance level of <span class="math inline">\(5 \times 10^-8\)</span>. Note that in case of a logistic regression model, a volcano plot should use the log-odds ratio (that is, the regression coefficient) on the <span class="math inline">\(x\)</span> axis. Using odds ratios would produce an asymmetric plot, as odds ratios are bounded below by zero, but are unbounded above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">plot</span>(crude[, .(beta, <span class="sc">-</span><span class="fu">log10</span>(p.value))], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, </span>
<span id="cb8-2"><a href="#cb8-2"></a>      <span class="at">main =</span> <span class="st">"Volcano plot"</span>, </span>
<span id="cb8-3"><a href="#cb8-3"></a>      <span class="at">xlab =</span> <span class="st">"Coefficient"</span>, </span>
<span id="cb8-4"><a href="#cb8-4"></a>      <span class="at">ylab =</span> <span class="st">"-log10(p-value)"</span>)</span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>) <span class="co"># genome-wide significance threshold</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A Manhattan plot is better suited to visualise the ordering of <code>SNP</code>s along the chromosome, and it can show that hits generally do not appear alone, but they form spikes because of linkage disequilibrium (<code>SNP</code>s that are closer together have more chances of being transmitted together).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">plot</span>(<span class="sc">-</span><span class="fu">log10</span>(crude<span class="sc">$</span>p.value), </span>
<span id="cb9-2"><a href="#cb9-2"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, </span>
<span id="cb9-3"><a href="#cb9-3"></a>     <span class="at">main =</span> <span class="st">"Manhattan plot"</span>, </span>
<span id="cb9-4"><a href="#cb9-4"></a>     <span class="at">xlab =</span> <span class="st">"Locus on chromosome 21"</span>, </span>
<span id="cb9-5"><a href="#cb9-5"></a>     <span class="at">ylab =</span> <span class="st">"-log10(p-value)"</span></span>
<span id="cb9-6"><a href="#cb9-6"></a>     )</span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>) <span class="co"># genome-wide significance threshold</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Note that in this particular example, we did not have genetic positions of each <code>SNP</code>: if we did, we should use them to better represent the spread of <code>SNP</code>s on the chromosome. Also, this dataset contains only one chromosome: if we had more than one, it would be better to differentiate chromosomes by colour, to make it easier to pinpoint where the hits are.</p>
</section><section id="genetic-risk-scores" class="level1"><h1>Genetic Risk Scores</h1>
<p>The <code>SNP</code>s that are most strongly associated with a specific trait or disease outcome can be used to build a genetic risk score (or polygenic risk score). A threshold on the p-value for such scores does not need to be as strict as the genome-wide significance threshold: often <span class="math inline">\(10^{-5}\)</span> or <span class="math inline">\(10^{-4}\)</span> are used, and in some cases those can be relaxed even further.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># subset of snps to enter the genetic risk score</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>chrom21.grs <span class="ot">&lt;-</span> chrom21[, .SD, .SDcols <span class="ot">=</span> crude[p.value <span class="sc">&lt;</span> <span class="fl">1e-5</span>]<span class="sc">$</span>snp]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can build a simple genetic risk score for each patient simply by summing the allele counts across all the associated <code>SNP</code>s (<em>unweighted score</em>). To take the direction of the effect into consideration, we switch the sign for the <code>SNP</code>s that have an inverse association: we use function <code><a href="https://rdrr.io/r/base/sign.html">sign()</a></code> for this purpose, which returns <span class="math inline">\(1\)</span> for positive elements, and <span class="math inline">\(-1\)</span> for negative elements.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>snps.grs <span class="ot">&lt;-</span> crude[p.value <span class="sc">&lt;</span> <span class="fl">1e-5</span>] </span>
<span id="cb11-2"><a href="#cb11-2"></a>chrom21.grs <span class="ot">&lt;-</span> chrom21[, .SD, .SDcols <span class="ot">=</span> crude[p.value <span class="sc">&lt;</span> <span class="fl">1e-5</span>]<span class="sc">$</span>snp] </span>
<span id="cb11-3"><a href="#cb11-3"></a>unweighted.score <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(chrom21.grs) <span class="sc">%*%</span> <span class="fu">sign</span>(snps.grs<span class="sc">$</span>beta) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once a score is built, it can be considered as an additional predictor and tested for association with the outcome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>mod.unweighted <span class="ot">&lt;-</span> <span class="fu">lm</span>(chrom21<span class="sc">$</span>outcome <span class="sc">~</span> unweighted.score)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">kable</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(mod.unweighted)), </span>
<span id="cb12-3"><a href="#cb12-3"></a>      <span class="at">caption =</span> <span class="st">"Summary Statistics with Unweighted Score"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb12-4"><a href="#cb12-4"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Summary Statistics with Unweighted Score</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> Estimate </th>
   <th style="text-align:right;"> Std. Error </th>
   <th style="text-align:right;"> t value </th>
   <th style="text-align:right;"> Pr(&gt;|t|) </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> (Intercept) </td>
   <td style="text-align:right;"> -0.0403636 </td>
   <td style="text-align:right;"> 0.0716479 </td>
   <td style="text-align:right;"> -0.5633603 </td>
   <td style="text-align:right;"> 0.5738238 </td>
  </tr>
<tr>
<td style="text-align:left;"> unweighted.score </td>
   <td style="text-align:right;"> 0.0481986 </td>
   <td style="text-align:right;"> 0.0051622 </td>
   <td style="text-align:right;"> 9.3368699 </td>
   <td style="text-align:right;"> 0.0000000 </td>
  </tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">summary</span>(mod.unweighted)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = chrom21$outcome ~ unweighted.score)

Residuals:
     Min       1Q   Median       3Q      Max 
-2.70347 -0.67316 -0.03193  0.71808  2.38092 

Coefficients:
                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      -0.040364   0.071648  -0.563    0.574    
unweighted.score  0.048199   0.005162   9.337   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.9733 on 199 degrees of freedom
Multiple R-squared:  0.3046,    Adjusted R-squared:  0.3011 
F-statistic: 87.18 on 1 and 199 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>More commonly, however, the <code>SNP</code>s that enter a score are weighted by their regression coefficient, so that not only the direction of the effect is taken into consideration, but also so that loci with larger effect contribute more to the final score. Whenever people refer to genetic (or polygenic) risk scores, it can be assumed that they refer to the weighted version.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>weighted.score <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(chrom21.grs) <span class="sc">%*%</span> snps.grs<span class="sc">$</span>beta </span>
<span id="cb15-2"><a href="#cb15-2"></a>mod.weighted <span class="ot">&lt;-</span> <span class="fu">lm</span>(chrom21<span class="sc">$</span>outcome <span class="sc">~</span> weighted.score)</span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="fu">kable</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(mod.weighted)), </span>
<span id="cb15-4"><a href="#cb15-4"></a>      <span class="at">caption =</span> <span class="st">"Summary Statistics with Weighted Score"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb15-5"><a href="#cb15-5"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Summary Statistics with Weighted Score</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> Estimate </th>
   <th style="text-align:right;"> Std. Error </th>
   <th style="text-align:right;"> t value </th>
   <th style="text-align:right;"> Pr(&gt;|t|) </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> (Intercept) </td>
   <td style="text-align:right;"> -0.0609842 </td>
   <td style="text-align:right;"> 0.0701831 </td>
   <td style="text-align:right;"> -0.8689291 </td>
   <td style="text-align:right;"> 0.3859327 </td>
  </tr>
<tr>
<td style="text-align:left;"> weighted.score </td>
   <td style="text-align:right;"> 0.0735971 </td>
   <td style="text-align:right;"> 0.0076163 </td>
   <td style="text-align:right;"> 9.6630505 </td>
   <td style="text-align:right;"> 0.0000000 </td>
  </tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="fu">summary</span>(mod.weighted)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = chrom21$outcome ~ weighted.score)

Residuals:
     Min       1Q   Median       3Q      Max 
-2.64046 -0.66038 -0.03924  0.69016  2.34576 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    -0.060984   0.070183  -0.869    0.386    
weighted.score  0.073597   0.007616   9.663   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.963 on 199 degrees of freedom
Multiple R-squared:  0.3194,    Adjusted R-squared:  0.3159 
F-statistic: 93.37 on 1 and 199 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The weighted score seems to perform marginally better, not only as measured by the Wald test p-value, but also by looking at the adjusted <span class="math inline">\(R^2 = 0.316\)</span> for the weighted score versus <span class="math inline">\(0.301\)</span> for the unweighted score. This tells us that the weighted score explains a bit more of the variance in the outcome.</p>
<p>Note that we cannot compare the regression coefficients directly, as the two scores are scaled differently: either we should have standardised the two scores to have the same standard deviation before fitting the models (as we did in <strong>Lab 2</strong>), or we correct the regression coefficients by multiplying each of them by the standard deviation of their corresponding predictor. Here we follow the latter approach, so the standardised regression coefficients become:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># standardised regression coefficients </span></span>
<span id="cb18-2"><a href="#cb18-2"></a>beta.unweighted.std <span class="ot">&lt;-</span> <span class="fu">coef</span>(mod.unweighted)[<span class="dv">2</span>] <span class="sc">*</span> <span class="fu">sd</span>(unweighted.score) </span>
<span id="cb18-3"><a href="#cb18-3"></a>beta.weighted.std <span class="ot">&lt;-</span> <span class="fu">coef</span>(mod.weighted)[<span class="dv">2</span>] <span class="sc">*</span> <span class="fu">sd</span>(weighted.score)</span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="fu">kable</span>(<span class="fu">round</span>(<span class="fu">c</span>(beta.unweighted.std, beta.weighted.std), <span class="dv">3</span>), </span>
<span id="cb18-5"><a href="#cb18-5"></a>      <span class="at">caption =</span> <span class="st">"Standardised Regression Coefficients"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb18-6"><a href="#cb18-6"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Standardised Regression Coefficients</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> x </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> unweighted.score </td>
   <td style="text-align:right;"> 0.643 </td>
  </tr>
<tr>
<td style="text-align:left;"> weighted.score </td>
   <td style="text-align:right;"> 0.658 </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>Also according to this criterion, it appears that the weighted score is a marginally better estimate of the genetic risk for the quantitative trait under study.</p>
<p>What would happen if we did not filter the <code>SNP</code>s that contribute to the genetic risk score? This would consider all the <code>SNP</code>s in the dataset as being related to the outcome under study: in some cases, it is not reasonable to expect that a score computed in this way would perform particularly well. However, complex traits (such as height) may be better explained by a large number of small effects spread throughout the genome than by a reduced set of most strongly associated <code>SNP</code>s.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co"># ensure that the ordering of snps ts respected</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="fu">stopifnot</span>(<span class="fu">colnames</span>(chrom21)[<span class="sc">-</span><span class="dv">1</span>] <span class="sc">==</span> crude<span class="sc">$</span>snp) </span>
<span id="cb19-3"><a href="#cb19-3"></a>all.score <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(chrom21[, <span class="sc">-</span><span class="dv">1</span>]) <span class="sc">%*%</span> crude<span class="sc">$</span>beta</span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co"># standardize the score to allow for direct comparison of effect sizes </span></span>
<span id="cb19-5"><a href="#cb19-5"></a>all.score <span class="ot">&lt;-</span> all.score <span class="sc">/</span> <span class="fu">sd</span>(all.score) </span>
<span id="cb19-6"><a href="#cb19-6"></a>mod.all <span class="ot">&lt;-</span> <span class="fu">lm</span>(chrom21<span class="sc">$</span>outcome <span class="sc">~</span> all.score)</span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="fu">kable</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(mod.all)), </span>
<span id="cb19-8"><a href="#cb19-8"></a>      <span class="at">caption =</span> <span class="st">"Summary Statistics with all SNPs"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb19-9"><a href="#cb19-9"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Summary Statistics with all SNPs</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> Estimate </th>
   <th style="text-align:right;"> Std. Error </th>
   <th style="text-align:right;"> t value </th>
   <th style="text-align:right;"> Pr(&gt;|t|) </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> (Intercept) </td>
   <td style="text-align:right;"> 0.0919377 </td>
   <td style="text-align:right;"> 0.0500695 </td>
   <td style="text-align:right;"> 1.83620 </td>
   <td style="text-align:right;"> 0.0678201 </td>
  </tr>
<tr>
<td style="text-align:left;"> all.score </td>
   <td style="text-align:right;"> 0.9520540 </td>
   <td style="text-align:right;"> 0.0475112 </td>
   <td style="text-align:right;"> 20.03851 </td>
   <td style="text-align:right;"> 0.0000000 </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>In this specific instance, using all <code>SNP</code>s produces a significant improvement in the fit of the model (<span class="math inline">\(R^2\)</span> is <span class="math inline">\(0.667\)</span>), as well as in effect size (<span class="math inline">\(0.952\)</span>). However, remember that the true assessment of performance of a genetic risk score, as for any other predictor, can be done only on withdrawn data (on a new set of observations or within a cross-validation setting).</p>
</section><section id="predicting-with-scores" class="level1"><h1>Predicting with scores</h1>
<p>Let us consider the case of a single training-test split, which can be created by using function <code>createDataPartition()</code> from the <code>caret</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>) </span>
<span id="cb20-2"><a href="#cb20-2"></a>train.idx <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(chrom21<span class="sc">$</span>outcome, <span class="at">p =</span> <span class="fl">0.7</span>)<span class="sc">$</span>Resample1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will use the training set for two purposes:</p>
<ol type="1">
<li>Run association tests for all the <code>SNP</code>s considered individually, and use the summary statistics to develop a genetic risk score.</li>
<li>Learn the regression coefficients for a model that uses the score.</li>
</ol>
<p>After this, we will be able to build genetic risk scores for the observations in the test set and make a prediction of the outcome.</p>
<p>We start by testing each <code>SNP</code> for association with the outcome: this uses the same approach developed earlier, but this time we restrict the data to the training set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>chrom21.train <span class="ot">&lt;-</span> chrom21[train.idx] </span>
<span id="cb21-2"><a href="#cb21-2"></a>crude <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="cf">for</span> (snp <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">ncol</span>(chrom21.train)) { </span>
<span id="cb21-4"><a href="#cb21-4"></a>  crude <span class="ot">&lt;-</span> <span class="fu">rbind</span>(crude,</span>
<span id="cb21-5"><a href="#cb21-5"></a>                 <span class="fu">univ.lm.test</span>(chrom21.train[[snp]], chrom21.train[[<span class="dv">1</span>]])) </span>
<span id="cb21-6"><a href="#cb21-6"></a>  }</span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="co"># add column of snp identifiers</span></span>
<span id="cb21-8"><a href="#cb21-8"></a>crude <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">snp=</span><span class="fu">colnames</span>(chrom21.train)[<span class="sc">-</span><span class="dv">1</span>], crude)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can create a weighted score by using the top <code>SNP</code>s. Note that we do it for all samples, also those in the test set: this is fine, we are not using the outcome variable, but just summing up the columns of the <code>SNP</code> matrix weighted by the coefficients learnt on the training set. Having a score for all samples will make it easier further on when testing the performance of the score.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>snps.grs <span class="ot">&lt;-</span> crude[p.value <span class="sc">&lt;</span> <span class="fl">1e-5</span>]</span>
<span id="cb22-2"><a href="#cb22-2"></a>chrom21.grs <span class="ot">&lt;-</span> chrom21[, .SD, .SDcols <span class="ot">=</span> crude[p.value <span class="sc">&lt;</span> <span class="fl">1e-5</span>]<span class="sc">$</span>snp] </span>
<span id="cb22-3"><a href="#cb22-3"></a>weighted.score <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(chrom21.grs) <span class="sc">%*%</span> snps.grs<span class="sc">$</span>beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is time to fit a model with the score variable: here it is important to use only the training set (defined by the subset option), as we are using the outcome to learn the regression coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>model.score <span class="ot">&lt;-</span> <span class="fu">lm</span>(outcome <span class="sc">~</span> weighted.score, <span class="at">data=</span>chrom21, <span class="at">subset=</span>train.idx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are finally ready to predict the observations in the test set. Given that the <code>weighted.score</code> is outside of the <code>chrom21</code> dataset, the easiest way to achieve this, is to predict all observations and then discard those that are in the training set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(model.score, <span class="at">newdata=</span>chrom21)[<span class="sc">-</span>train.idx]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As a measure of performance we can compute the correlation coefficient between predicted and observed outcome (restricted to samples in the test set), or the square of that (corresponding to the <span class="math inline">\(R^2\)</span>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">round</span>(<span class="fu">cor</span>(pred, chrom21<span class="sc">$</span>outcome[<span class="sc">-</span>train.idx]), <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.44</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">round</span>(<span class="fu">cor</span>(pred, chrom21<span class="sc">$</span>outcome[<span class="sc">-</span>train.idx])<span class="sc">^</span><span class="dv">2</span>, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.193</code></pre>
</div>
</div>
</section><section id="gwas-meta-analysis" class="level1"><h1>GWAS meta-analysis</h1>
<p>When genome-wide association studies of the same phenotype are performed in multiple independent studies, it is common to merge the results by performing a meta-analysis. ’This has the benefit of increasing the statistical power and reducing false-positive findings (type I error rate). Again, this type of analysis is most often done through specialised software (such as <code>plink</code>, <code>METAL</code>, <code>GWAMA</code>, etc) which can perform several checks on the input files and produce helpful diagnostics.</p>
<p>Before being able to perform a meta-analysis, the summary statistics from the studies need to be harmonised, to ensure that the effect sizes reported refer to the same allele. Thus, if two studies used a different effect allele, the sign for the effect size of one of them should be switched.</p>
<p>Files <code>OPN_sub_study1.txt</code> and <code>OPN_sub_study2.txt</code> contain a small subset of results from two different GWASs run on Osteopontin (OPN) a protein that can be measured in blood serum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co"># Use fread() to load data. It does not convert strings to factors </span></span>
<span id="cb29-2"><a href="#cb29-2"></a>gwas1 <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/OPN_sub_study1.txt"</span>) </span>
<span id="cb29-3"><a href="#cb29-3"></a>gwas2 <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/OPN_sub_study2.txt"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We need to harmonise the two datasets: in this specific case it is enough to order them by chromosome and position, but we need to ensure that the <code>SNP</code> identifiers correspond.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>gwas1 <span class="ot">&lt;-</span> gwas1[rsid <span class="sc">%in%</span> gwas2<span class="sc">$</span>rsid]</span>
<span id="cb30-2"><a href="#cb30-2"></a>gwas2 <span class="ot">&lt;-</span> gwas2[rsid <span class="sc">%in%</span> gwas1<span class="sc">$</span>rsid] </span>
<span id="cb30-3"><a href="#cb30-3"></a></span>
<span id="cb30-4"><a href="#cb30-4"></a><span class="co"># order by chromosome and position </span></span>
<span id="cb30-5"><a href="#cb30-5"></a>gwas1 <span class="ot">&lt;-</span> gwas1[<span class="fu">order</span>(chromosome, position)] </span>
<span id="cb30-6"><a href="#cb30-6"></a>gwas2 <span class="ot">&lt;-</span> gwas2[<span class="fu">order</span>(chromosome, position)]</span>
<span id="cb30-7"><a href="#cb30-7"></a><span class="fu">all.equal</span>(gwas1<span class="sc">$</span>rsid, gwas2<span class="sc">$</span>rsid) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Ignore the differences in position between the two studies: they may have been carried out using different genome builds (the relative order of <code>SNP</code>s is unaffected). What matters is that <code>SNP</code> names and both alleles match, and that both studies report the effect size on the same allele. In this example, both studies considered Al to be the effect allele, but different studies may have different conventions, and they need to be harmonised as well in case of discrepancies.</p>
<p>We start by finding which <code>SNP</code>s have both alleles matching, and which would match if the alleles were flipped.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>both.ok <span class="ot">&lt;-</span> gwas1<span class="sc">$</span>A1 <span class="sc">==</span> gwas2<span class="sc">$</span>A1 <span class="sc">&amp;</span> gwas1<span class="sc">$</span>A2 <span class="sc">==</span> gwas2<span class="sc">$</span>A2 </span>
<span id="cb32-2"><a href="#cb32-2"></a>flipped <span class="ot">&lt;-</span> gwas1<span class="sc">$</span>A1 <span class="sc">==</span> gwas2<span class="sc">$</span>A2 <span class="sc">&amp;</span> gwas1<span class="sc">$</span>A2 <span class="sc">==</span> gwas2<span class="sc">$</span>A1</span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="fu">kable</span>(<span class="fu">table</span>(both.ok, flipped), </span>
<span id="cb32-4"><a href="#cb32-4"></a>      <span class="at">caption =</span> <span class="st">"SNPs with Alleles Matching"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb32-5"><a href="#cb32-5"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>SNPs with Alleles Matching</caption>
 <thead><tr>
<th style="text-align:left;">   </th>
   <th style="text-align:right;"> FALSE </th>
   <th style="text-align:right;"> TRUE </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:left;"> FALSE </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 25 </td>
  </tr>
<tr>
<td style="text-align:left;"> TRUE </td>
   <td style="text-align:right;"> 27 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>In this case, we can see that the alleles either match already, or they match after flipping them. If both alleles do not match even after swapping them, that <code>SNP</code> cannot be meta-analysed. This could be caused by a genotyping or imputation error, or because of some other discrepancy in the GWAS pipelines adopted by the different studies.</p>
<p>The effect sizes of the <code>SNP</code>s which we identified to have alleles flipped need to have their direction of effect swapped in one of the studies before entering the meta-analysis. Here we swap the sign for the second study, but this is an arbitrary choice.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>beta1 <span class="ot">&lt;-</span> gwas1<span class="sc">$</span>Effect_A1</span>
<span id="cb33-2"><a href="#cb33-2"></a>beta2 <span class="ot">&lt;-</span> gwas2<span class="sc">$</span>Effect_A1</span>
<span id="cb33-3"><a href="#cb33-3"></a>beta2[flipped] <span class="ot">&lt;-</span> <span class="sc">-</span>beta2[flipped]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally we are ready to perform a fixed effect meta-analysis by using inverse variance weighting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>weight.gwas1 <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> gwas1<span class="sc">$</span>StdErr_A1<span class="sc">^</span><span class="dv">2</span> </span>
<span id="cb34-2"><a href="#cb34-2"></a>weight.gwas2 <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> gwas2<span class="sc">$</span>StdErr_A1<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By looking at the weights assigned to the two studies, it appears that in general the second study is more powered.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="fu">kable</span>(<span class="fu">t</span>(<span class="fu">head</span>(weight.gwas1)), </span>
<span id="cb35-2"><a href="#cb35-2"></a>      <span class="at">caption =</span> <span class="st">"Weight of 1st GWAS"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb35-3"><a href="#cb35-3"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Weight of 1st GWAS</caption>
<tbody><tr>
<td style="text-align:right;"> 929.5062 </td>
   <td style="text-align:right;"> 734.4247 </td>
   <td style="text-align:right;"> 730.4602 </td>
   <td style="text-align:right;"> 935.2 </td>
   <td style="text-align:right;"> 935.2 </td>
   <td style="text-align:right;"> 929.5062 </td>
  </tr></tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="fu">kable</span>(<span class="fu">t</span>(<span class="fu">head</span>(weight.gwas2)), </span>
<span id="cb36-2"><a href="#cb36-2"></a>      <span class="at">caption =</span> <span class="st">"Weight of 2nd GWAS"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb36-3"><a href="#cb36-3"></a>  <span class="fu">kable_styling</span>(<span class="at">full_width =</span> F, <span class="at">position =</span> <span class="st">"center"</span>, <span class="at">latex_options =</span> <span class="st">"hold_position"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>Weight of 2nd GWAS</caption>
<tbody><tr>
<td style="text-align:right;"> 1339.1 </td>
   <td style="text-align:right;"> 1104.129 </td>
   <td style="text-align:right;"> 1118.638 </td>
   <td style="text-align:right;"> 1401.177 </td>
   <td style="text-align:right;"> 1379.63 </td>
   <td style="text-align:right;"> 1348.865 </td>
  </tr></tbody>
</table>
</div>
</div>
<p>The compuation of the meta-analysis effect size is a weighted sum of the effect sizes from each study, weighted according to the weight just derived.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>beta.ma <span class="ot">&lt;-</span> (weight.gwas1 <span class="sc">*</span> beta1 <span class="sc">+</span> weight.gwas2 <span class="sc">*</span> beta2) <span class="sc">/</span> (weight.gwas1 <span class="sc">+</span> weight.gwas2) </span>
<span id="cb37-2"><a href="#cb37-2"></a>se.ma <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="dv">1</span> <span class="sc">/</span> (weight.gwas1 <span class="sc">+</span> weight.gwas2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that by combining the two studies through a meta-analysis we have increased the power. This can be seen by comparing the p-values of the first study (which has them available already) to the meta-analysis p-values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>pval.ma <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="fu">abs</span>(beta.ma <span class="sc">/</span> se.ma), <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="fu">plot</span>(<span class="sc">-</span><span class="fu">log10</span>(gwas1<span class="sc">$</span>P.value), <span class="sc">-</span><span class="fu">log10</span>(pval.ma), </span>
<span id="cb38-3"><a href="#cb38-3"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">16</span>), </span>
<span id="cb38-4"><a href="#cb38-4"></a>     <span class="at">xlab =</span> <span class="st">"p-values from gwas1"</span>, </span>
<span id="cb38-5"><a href="#cb38-5"></a>     <span class="at">ylab =</span> <span class="st">"p-values from meta-analysis"</span>,</span>
<span id="cb38-6"><a href="#cb38-6"></a>     <span class="at">main =</span> <span class="st">"P-values Comparison between GWAS and Meta-Analysis"</span></span>
<span id="cb38-7"><a href="#cb38-7"></a>     )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Lab_5_Using_Genetic_Data_to_Predict_Outcomes_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div style="page-break-after: always;"></div>
</section><section id="practical-exercises" class="level1"><h1>Practical Exercises</h1>
<section id="question-1." class="level2"><h2 class="anchored" data-anchor-id="question-1.">Question 1.</h2>
<p>Using file <code>chrom21.csv</code> (available on Learn):</p>
<ul>
<li>Impute the missing values in the <code>SNP</code>s using mean imputation method.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Set the random seed to <span class="math inline">\(1\)</span> and create <span class="math inline">\(5\)</span> cross-validation folds.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Within each training fold, perform an association study for all <code>SNP</code>s in the dataset (hint: store the data table of results as an element in a list) and report the most associated <code>SNP</code>. (answer: snp969, snp969, snp969, snp969, snp969).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Within each training fold, build a weighted genetic risk score that uses <code>SNP</code>s filtered at p-value <span class="math inline">\(&lt;10^{-4}\)</span>, as well as a weighted genetic risk score that uses all available <code>SNP</code>s (hint: build the scores for all observations, also those in the test set). Build the cross-validation folds for filtered <code>SNP</code>s and all <code>SNP</code>s.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Within each training fold, fit a linear regression models for each score. (hint: store each regression object as an element in a list).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Use the two genetic risk scores to predict the outcome in each of the test folds and compute the correlation coefficient between the observed outcome and the predicted outcome. Report the average correlation over all folds. (answer: <span class="math inline">\(0.556\)</span>, <span class="math inline">\(0.739\)</span>).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="question-2" class="level2"><h2 class="anchored" data-anchor-id="question-2">Question 2</h2>
<p>Using file <code>dpt2.txt</code> (available on Learn):</p>
<ul>
<li>Sort the data table in increasing order of chromosome number and base pairs (column <code>pos</code>).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Using function <code><a href="https://rdrr.io/r/base/numeric.html">as.numeric()</a></code>, convert the <code>pos</code> column from integer to numeric (to avoid overflow). Add column <code>cum.pos</code> to the data table to contain the cumulative sum of the positions (use function <code><a href="https://rdrr.io/r/base/cumsum.html">cumsum()</a></code>), and assert that no <code>NA</code>s were produced.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Create a Manhattan plot that orders the <code>SNP</code>s according to their cumulative position along the genome</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Redo the plot by using alternating colours for adjacent chromosomes, and add a horizontal line corresponding to the standard genome-wide significance threshold.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Produce a volcano plot for this set of results, using a different colour for <code>SNP</code>s that have minor allele frequency (MAF) <span class="math inline">\(&lt; 5\%\)</span> Volcano plot:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="co">#Answer in this chunk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section></main><!-- /main column --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>